source/fab/metrics.py

```diff
@@ -0,0 +1,214 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+"""
+A module for recording and summarising metrics, with the following concepts:
+
+init
+    create pipes
+    create reading process - daemonic, exits with main process
+
+send
+    group, name, value -> reading process
+    overwrites any previous value for group[name]
+
+reading process
+    creates and add to metrics dict
+    finishes -> send whole lot down a summary pipe and close
+
+stop
+    closes pipes & process
+    return metrics from summary pipe
+
+"""
+import datetime
+import json
+import logging
+from collections import defaultdict
+from multiprocessing import Process, Pipe
+from multiprocessing.connection import Connection
+from pathlib import Path
+from typing import Optional, Dict
+
+logger = logging.getLogger('fab')
+
+# the pipe for individual metrics
+_metric_recv_conn: Optional[Connection] = None
+_metric_send_conn: Optional[Connection] = None
+
+# the pipe connection for the final collated metrics
+_collated_metrics_recv_conn: Optional[Connection] = None
+_collated_metrics_send_conn: Optional[Connection] = None
+
+# the process which receives individual metrics
+_metric_recv_process: Optional[Process] = None
+
+
+def init_metrics():
+    """
+    Create the pipe for sending metrics, the process to read them, and another pipe to push the final collated data.
+
+    Only one call to init_metrics can be called before calling stop_metrics.
+
+    """
+    global _metric_recv_conn, _metric_send_conn
+    global _collated_metrics_recv_conn, _collated_metrics_send_conn
+    global _metric_recv_process
+
+    if any([_metric_recv_conn, _metric_send_conn, _collated_metrics_recv_conn, _collated_metrics_send_conn]):
+        raise ConnectionError('Metrics already initialised. Only one concurrent user of init_metrics is expected.')
+
+    # the pipe connections for individual metrics
+    _metric_recv_conn, _metric_send_conn = Pipe(duplex=False)
+
+    # the pipe connections for the collated metrics, once they've stopped being sent
+    _collated_metrics_recv_conn, _collated_metrics_send_conn = Pipe(duplex=False)
+
+    # start the receiving process
+    _metric_recv_process = Process(
+        target=_read_metric,
+        daemon=True,  # todo: test this thoroughly, manually
+    )
+    _metric_recv_process.start()
+
+
+def _read_metric():
```

Comment:
Normally the pattern here would be to pass the necessary pipe endpoints as context to the process, rather than using globals. I think the `args` argument of the `Process` constructor allows for this.

Reply:
Agreed. We've raised a [ticket to consider metrics](https://github.com/metomi/fab/projects/8#card-81420757), which might make this redundant.

Reply:
I've updated the note with this point.